src/ddmd/expressionsem.d

```diff
@@ -8771,6 +8808,100 @@ Expression binSemanticProp(BinExp e, Scope* sc)
     return null;
 }
 
+/********************************************************************************
+ * Helper function to consolidate common code for visit(BinAssignExp)
+ * visit(PowAssignExp), and visit(CatAssignExp).  It rewrites expressions
+ * of the form, for example, `s.x += 1` to `s.x(s.x() += 1)` if s.x is
+ * a getter and setter `@property`.
+ * Params:
+ *      e = the expression to rewrite
+ *      sc = the semantic scope
+ * Returns:
+ *      the rewritten expression if the procedure succeeds, an `ErrorExp` if the
+ *      and error is encountered, or `null` if `e.e1` is not a `DotIdExp`.
+ */
+Expression binAssignRewriteProp(BinAssignExp e, Scope* sc)
+{
+    // https://issues.dlang.org/show_bug.cgi?id=8006
+
+    if (e.e1.op == TOKdotid)
+    {
+        // Consider s.x += 1;
+        // We want to rewrite that expression to s.x(s.x() + 1);
+        // However, resolvePropertiesX will change the type of s.x() to void,
```

Comment:
Thanks.  Looking at this after sleeping on it, I sense that what I am see is infact memory that is being shared between both `s.x()` and `s.x(1)` calls.  When the latter gets rewritten, it changes the type of the former also (oops!).

I'm just having a look now.  I think this is close to being where it should be anyway, just a few nits.